crates/polars-parquet/src/parquet/encoding/hybrid_rle/encoder.rs

```diff
@@ -3,98 +3,232 @@ use std::io::Write;
-    let mut container = [0; 10];
-    let used = uleb128::encode(header, &mut container);
-    writer.write_all(&container[..used])?;
+const U32_BLOCK_LEN: usize = 32;
 
-    bitpacked_encode_u32(writer, iterator, num_bits as usize)?;
+impl Encoder<u32> for u32 {
+    fn bitpacked_encode<W: Write, I: Iterator<Item = u32>>(
+        writer: &mut W,
+        mut iterator: I,
+        num_bits: usize,
+    ) -> std::io::Result<()> {
+        // the length of the iterator.
+        let length = iterator.size_hint().1.unwrap();
+
+        let mut header = ceil8(length) as u64;
+        header <<= 1;
+        header |= 1; // it is bitpacked => first bit is set
+        let mut container = [0; 10];
+        let used = uleb128::encode(header, &mut container);
+        writer.write_all(&container[..used])?;
+
+        let chunks = length / U32_BLOCK_LEN;
+        let remainder = length - chunks * U32_BLOCK_LEN;
+        let mut buffer = [0u32; U32_BLOCK_LEN];
+
+        // simplified from ceil8(U32_BLOCK_LEN * num_bits) since U32_BLOCK_LEN = 32
+        let compressed_chunk_size = 4 * num_bits;
+
+        for _ in 0..chunks {
+            iterator
+                .by_ref()
+                .take(U32_BLOCK_LEN)
+                .zip(buffer.iter_mut())
+                .for_each(|(item, buf)| *buf = item);
+
+            let mut packed = [0u8; 4 * U32_BLOCK_LEN];
+            bitpacked::encode_pack::<u32>(&buffer, num_bits, packed.as_mut());
+            writer.write_all(&packed[..compressed_chunk_size])?;
+        }
+
+        if remainder != 0 {
+            // Must be careful here to ensure we write a multiple of `num_bits`
+            // (the bit width) to align with the spec. Some readers also rely on
+            // this - see https://github.com/pola-rs/polars/pull/13883.
+
+            // this is ceil8(remainder * num_bits), but we ensure the output is a
+            // multiple of num_bits by rewriting it as ceil8(remainder) * num_bits
+            let compressed_remainder_size = ceil8(remainder) * num_bits;
+            iterator
+                .by_ref()
+                .take(remainder)
+                .zip(buffer.iter_mut())
+                .for_each(|(item, buf)| *buf = item);
+
+            let mut packed = [0u8; 4 * U32_BLOCK_LEN];
+            // No need to zero rest of buffer because remainder is either:
+            // * Multiple of 8: We pad non-terminal literal runs to have a
+            //   multiple of 8 values. Once compressed, the data will end on
+            //   clean byte boundaries and packed[..compressed_remainder_size]
+            //   will include only the remainder values and nothing extra.
+            // * Final run: Extra values from buffer will be included in
+            //   packed[..compressed_remainder_size] but ignored when decoding
+            //   because they extend beyond known column length
+            bitpacked::encode_pack(&buffer, num_bits, packed.as_mut());
```

Comment:
In my original PR, I changed this to `&buffer[..remainder]` because I was worried about other data being included from the buffer (same buffer is used before reaching this branch). 

I now realize that this is a non-issue for the reasons described in my comment and changed it back to `&buffer`. This theoretically avoids an extra memory allocation in `bitpacked::encode_pack`. Unfortunately, the current `bitpacked::encode_pack` always copies input data to a new array because it compares the input slice length to `T::Packed::LENGTH` instead of `T::Unpacked::LENGTH`. I'll fix that and optimize bit-packing in a separate PR.